### PR TITLE
feat(vite-plugin-nitro): prerender content dir routes

### LINKED ISSUE #820: Prerender content routes based on files in directory
### Which scope/s are relevant/related to the feature request?

platform

### Information

As described in the [Docs](https://analogjs.org/docs/features/server/static-site-generation), static site generation is currently done by adding routes to the `prerender` config in `vite.config.ts` like this:

```ts
import { defineConfig } from 'vite';
import analog from '@analogjs/platform';

// https://vitejs.dev/config/
export default defineConfig(({ mode }) => ({
  plugins: [
    analog({
      static: true,
      prerender: {
        routes: async () => [
          '/',
          '/about',
          '/blog',
          '/blog/posts/2023-02-01-my-first-post',
        ],
      },
    }),
  ],
}));
```

To alway prerender new sites for e. g. new blog post articles, you have to remember to update this config to be sure the articles are pre-rendered.
A common use-case is to alway pre-render such kind of articles or to pre-render them when front-matter meta-data is defined.

I would suggest to allow adding directory configs to the array of pre-rendered routes which auto-magically prerenders new articles like this:

```ts
import { defineConfig } from 'vite';
import analog from '@analogjs/platform';

// https://vitejs.dev/config/
export default defineConfig(({ mode }) => ({
  plugins: [
    analog({
      static: true,
      prerender: {
        routes: async () => [
          '/',
          '/about',
          '/blog',
          // previeously
          '/blog/posts/2023-02-01-my-first-post',
          // new:
          {
            dir: '/blog/posts', // matches all routes created by file-based routing within // `src/content/blog/posts`
            filter: (metaData) => { // optional: allow to filter pre-rendered routes based on front-matter meta data
              return metaData.published !== 'false'
            }
          } 
        ],
      },
    }),
  ],
}));
```

### Describe any alternatives/workarounds you're currently using

_No response_

### I would be willing to submit a PR to fix this issue

- [x] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by d-koppenhagen**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/analogjs/analog/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] Other... Please describe:

## Which package are you modifying?

- [ ] vite-plugin-angular
- [x] vite-plugin-nitro
- [ ] astro-angular
- [ ] create-analog
- [ ] router
- [x] platform
- [ ] content
- [ ] nx-plugin
- [ ] trpc

## What is the current behavior?

Closes #820

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by brandonroberts** [comment]
@d-koppenhagen I think this is on the right track. Will you check on the build failures?

image

**Comment by d-koppenhagen** [comment]
@brandonroberts I refactored it a bit and added also some more documentation and fixed the tests. It should now work as expected.

**Comment by brandonroberts** [comment]
@d-koppenhagen there are lots of files changed that aren't necessary for this PR. Any idea what happened there?

**Comment by d-koppenhagen** [comment]
I had run prettier as a GitHub action was red since formatting was not ok.
It's a separate commit.
I can revert it and cleanup my commits a bit.

**Comment by d-koppenhagen** [comment]
I think a big part of the changed files are the ones of the blog demo app which I updated to reflect the feature here.

**Comment by d-koppenhagen** [comment]
@brandonroberts sorry it took a while but now the build and format etc. is green.

File: apps/docs-app/docs/features/server/static-site-generation.md [hunk 1]
```diff
@@ -4,6 +4,8 @@ Analog supports Static Site Generation when building for deployment. This includ
 
 ## Static Site Generation
 
+### from routes list
```
Comment (brandonroberts) [inline apps/docs-app/docs/features/server/static-site-generation.md:7]:
```suggestion
### From Routes List
```

File: apps/docs-app/docs/features/server/static-site-generation.md [hunk 2]
```diff
@@ -27,6 +29,50 @@ export default defineConfig(({ mode }) => ({
 }));
 ```
 
+### from content directory
```
Comment (brandonroberts) [inline apps/docs-app/docs/features/server/static-site-generation.md:32]:
```suggestion
### From Content Directory
```

File: apps/docs-app/docs/features/server/static-site-generation.md [hunk 3]
```diff
@@ -27,6 +29,50 @@ export default defineConfig(({ mode }) => ({
...
+
+You might want to prerender all routes that are the result of a rendered content directory.
+For example if you have a blog and all your articles are places as Markdown files in the `contents` directory.
+For such scenarios, you can add an object to the `routes` config to render everything within a directory.
+Keep in mind, that your directory structure may not be reflected 1:1 in your apps path.
+Therefore, you have to pass a `transform` function which maps the file paths to the URLs.
+The returning string should be the URL path in your app.
+Using `transform` allows you also, so filter out some routes by returning `false`.
```
Comment (brandonroberts) [inline apps/docs-app/docs/features/server/static-site-generation.md:40]:
```suggestion
Using `transform` allows you also filter out some routes by returning `false`.
```

File: apps/docs-app/docs/features/server/static-site-generation.md [hunk 4]
```diff
@@ -27,6 +29,50 @@ export default defineConfig(({ mode }) => ({
...
+This will not include them into the prerender process (e. g. files marked as `draft` in the frontmatter).
```
Comment (brandonroberts) [inline apps/docs-app/docs/features/server/static-site-generation.md:41]:
```suggestion
This does not include them in the prerender process, such as files marked as `draft` in the frontmatter.
```

**Review by brandonroberts — APPROVED** [review]

File: packages/vite-plugin-nitro/src/lib/options.ts [hunk 5]
```diff
@@ -31,3 +33,32 @@ export interface PrerenderOptions {
...
+/**
+ * @param path the path to the content file
+ * @param name the basename of the matching content file without the file extension
+ * @param extension the file extension
+ * @param frontmatter the frontmatter attributes extracted from the frontmatter section of the file
+ * @returns a string with the route should be returned (e. g. `/blog/<slug>`) or the value `false`, when the route should not be prerendered.
+ */
+export interface ContentFile {
+  path: string;
+  frontmatter: Record<string, any>;
```
Comment (brandonroberts) [inline packages/vite-plugin-nitro/src/lib/options.ts:61]: Noticed this on the second pass. There is a similar interface in the @analogjs/content package named `ContentFile`. I think we should rename this to `PrerenderContentFile` and change `frontmatter` to `attributes`.

**Review by brandonroberts — CHANGES_REQUESTED** [review]

File: packages/vite-plugin-nitro/src/lib/options.ts (hunk 5 above)
Comment (d-koppenhagen) [reply packages/vite-plugin-nitro/src/lib/options.ts]: Done :)

File: packages/platform/src/lib/options.ts [hunk 6]
```diff
@@ -33,3 +35,32 @@ export interface Options {
...
+}
+
+/**
+ * @param path the path to the content file
+ * @param name the basename of the matching content file without the file extension
+ * @param extension the file extension
+ * @param attributes the frontmatter attributes extracted from the frontmatter section of the file
+ * @returns a string with the route should be returned (e. g. `/blog/<slug>`) or the value `false`, when the route should not be prerendered.
+ */
+export interface ContentFile {
```
Comment (brandonroberts) [inline packages/platform/src/lib/options.ts:61]:
```suggestion
export interface PrerenderContentFile {
```

**Review by brandonroberts — CHANGES_REQUESTED** [review]

**Comment by brandonroberts** [comment]
Thanks @d-koppenhagen!

**Pull request merged**